src/crimson/osd/replicated_backend.cc

```diff
@@ -103,3 +112,1005 @@ void ReplicatedBackend::got_rep_op_reply(const MOSDRepOpReply& reply)
     }
   }
 }
+
+seastar::future<> ReplicatedBackend::recover_object(
+  const hobject_t& soid,
+  eversion_t need)
+{
+  logger().debug("{}: {}, {}", __func__, soid, need);
+  auto& wfor = recovering[soid];
+  return seastar::do_with(PushOp(), get_shards_to_push(soid),
+    [this, soid, need, &wfor](auto& pop, auto& shards) {
+    return [this, soid, need, &wfor] {
+      pg_missing_tracker_t local_missing = pg.get_local_missing();
+      if (local_missing.is_missing(soid)) {
+	PullOp po;
+	auto& pi = wfor.pi;
+	prepare_pull(po, pi, soid, need);
+	auto msg = make_message<MOSDPGPull>();
+	msg->from = pg.get_pg_whoami();
+	msg->set_priority(pg.get_recovery_op_priority());
+	msg->pgid = pg.get_pgid();
+	msg->map_epoch = pg.get_osdmap_epoch();
+	msg->min_epoch = pg.get_last_peering_reset();
+	std::vector<PullOp> pulls;
+	pulls.push_back(po);
+	msg->set_pulls(&pulls);
+	(void)shard_services.send_to_osd(pi.from.osd,
+					 std::move(msg),
+					 pg.get_osdmap_epoch());
+	return wfor.wait_for_pull();
```

Comment:
@athanatos I mean, here, I'm waiting on a future that are resolved when the pull operation is completed, I think this would indicate that the message is already sent, is this right?

Reply:
@athanatos Do you mean I can do something like: `return shared_services.send_to_osd().then([] {return wfor.wait_for_pull()})`?

Reply:
I'd really rather you chain that wait_for_pull after the send_to_osd.  Basically, the only time we want to drop a future is when we actually mean to spawn an independent task.

Reply:
Got it:-)